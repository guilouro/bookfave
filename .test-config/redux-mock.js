import React from 'react'
import { createStore, combineReducers, applyMiddleware } from 'redux'
import { render } from 'react-testing-library'
import { Provider } from 'react-redux'

export const mockStore = mock => {
  const reducers = {}

  Object.keys(mock).forEach(key => {
    reducers[key] = () => mock[key]
  })

  return createStore(combineReducers(reducers))
}

export const renderWithRedux = (component, store) =>
  render(<Provider store={store}>{component}</Provider>)
