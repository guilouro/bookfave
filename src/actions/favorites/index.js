import uuid from 'uuid/v4'

export const ADD = '@favorite/ADD'
export const REMOVE = '@favorite/REMOVE'
export const REMOVE_TAG = '@tag/REMOVE'
export const FILTER = '@tag/FILTER'

const parseTags = tags => {
  const obj = {}
  tags.split(' ').forEach(tag => {
    obj[tag.toLowerCase()] = true
  })

  return obj
}

export const addFavorite = ({ title, link, tags }) => ({
  type: ADD,
  payload: {
    id: uuid(),
    title,
    link,
    tags: parseTags(tags)
  }
})

export const removeFavorite = id => ({
  type: REMOVE,
  payload: { id }
})

// TODO: Maybe to create a Tags actions
export const removeTag = (favoriteId, tagName) => ({
  type: REMOVE_TAG,
  payload: { favoriteId, tagName }
})

export const filterByTag = tagName => ({
  type: FILTER,
  payload: { tagName }
})
