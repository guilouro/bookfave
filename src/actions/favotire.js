import uuid from 'uuid/v4'

const ADD = '@favorite/ADD'
const REMOVE = '@favorite/REMOVE'
const REMOVE_TAG = '@tag/REMOVE'

const parseTags = tags => {
  const obj = {}
  tags.split('').forEach(tag => {
    obj[tag] = true
  })

  return obj
}

export const addFavorite = ({ title, link, tags }) => ({
  type: ADD,
  payload: {
    id: uuid(),
    title,
    link,
    tags: parseTags(tags)
  }
})

export const removeFavorite = id => ({
  type: REMOVE,
  payload: { id }
})

// TODO: Maybe to create a Tags actions
export const removeTag = tagName => ({
  type: REMOVE_TAG,
  payload: { tagName }
})
