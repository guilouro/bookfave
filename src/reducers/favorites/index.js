const initialState = {
  entities: {},
  allIds: []
}

const favorites = (state = initialState, action) => {
  switch (action.type) {
    default:
      return state
  }
}

export default favorites
