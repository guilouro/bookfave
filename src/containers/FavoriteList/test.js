import React from 'react'
import { cleanup, fireEvent } from 'react-testing-library'
import FavoriteList from '.'
import { mockStore, renderWithRedux } from '../../../.test-config/redux-mock'
import { removeFavorite, filterByTag, removeTag } from '../../actions/favorites'

describe('<FavoriteList />', () => {
  afterEach(cleanup)
  describe('without data', () => {
    const store = mockStore({
      favorites: {
        filterTag: '',
        entities: {},
        ids: []
      }
    })

    it('Should render correctly', () => {
      const { container, queryByText } = renderWithRedux(
        <FavoriteList />,
        store
      )
      expect(queryByText(/We don't have any favorite/g)).toBeTruthy()
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('with data', () => {
    const store = mockStore({
      favorites: {
        filterTag: 'wallet',
        entities: {
          1: {
            id: 1,
            title: 'Lorem Ipsum',
            link: 'www.loremipsum.com',
            tags: {
              wallet: true
            }
          }
        },
        ids: [1]
      }
    })

    it('Should render correctly', () => {
      const { container, queryByText } = renderWithRedux(
        <FavoriteList />,
        store
      )

      expect(queryByText(/Lorem Ipsum/g)).toBeTruthy()
      expect(queryByText(/www.loremipsum.com/g)).toBeTruthy()
      expect(queryByText(/wallet/g)).toBeTruthy()
      expect(container.firstChild).toMatchSnapshot()
    })

    it('Should called onRemoveItem', () => {
      store.dispatch = jest.fn()
      const { queryByText } = renderWithRedux(<FavoriteList />, store)

      fireEvent.click(queryByText(/Delete/g))
      expect(store.dispatch).toBeCalledWith(removeFavorite(1))
    })

    it('Should called onFilterTag', () => {
      store.dispatch = jest.fn()
      const { queryByText } = renderWithRedux(<FavoriteList />, store)

      fireEvent.click(queryByText(/wallet/g))
      expect(store.dispatch).toBeCalledWith(filterByTag('wallet'))
    })

    it('Should called onRemoveTag', () => {
      store.dispatch = jest.fn()
      const { queryByText } = renderWithRedux(<FavoriteList />, store)

      fireEvent.click(queryByText(/âœ•/g))
      expect(store.dispatch).toBeCalledWith(removeTag(1, 'wallet'))
    })
  })
})
