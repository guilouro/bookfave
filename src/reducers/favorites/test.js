import reducer, { searchIdByTag } from '.'
import { ADD, REMOVE, REMOVE_TAG, FILTER } from '../../actions/favorite'

const initialState = {
  filterTag: '',
  entities: {},
  ids: []
}

describe('Reducers - favorite', () => {
  it('Should return initial state', () => {
    const state = reducer(undefined, {})
    expect(state).toEqual(initialState)
  })

  it('Should handle ADD', () => {
    const expected = {
      filterTag: '',
      entities: {
        1: {
          id: 1,
          title: 'Lorem Ipsum',
          link: 'www.loremipsum.com',
          tags: { wallet: true, credit: true }
        }
      },
      ids: [1]
    }

    const state = reducer(initialState, {
      type: ADD,
      payload: {
        id: 1,
        title: 'Lorem Ipsum',
        link: 'www.loremipsum.com',
        tags: { wallet: true, credit: true }
      }
    })

    expect(state).toEqual(expected)
  })

  it('Should handle REMOVE', () => {
    const initial = {
      entities: {
        1: {
          id: 1,
          title: 'Lorem Ipsum',
          link: 'www.loremipsum.com',
          tags: { wallet: true, credit: true }
        },
        5: {
          id: 5,
          title: 'Lorem Ipsum',
          link: 'www.loremipsum.com',
          tags: { wallet: true, credit: true }
        }
      },
      ids: [1, 5]
    }

    const expected = {
      entities: {
        1: {
          id: 1,
          title: 'Lorem Ipsum',
          link: 'www.loremipsum.com',
          tags: { wallet: true, credit: true }
        }
      },
      ids: [1]
    }

    const state = reducer(initial, {
      type: REMOVE,
      payload: { id: 5 }
    })

    expect(state).toEqual(expected)
  })

  it('Should handle REMOVE_TAG', () => {
    const initial = {
      entities: {
        1: {
          id: 1,
          title: 'Lorem Ipsum',
          link: 'www.loremipsum.com',
          tags: { wallet: true, credit: true }
        }
      },
      ids: [1]
    }

    const expected = {
      entities: {
        1: {
          id: 1,
          title: 'Lorem Ipsum',
          link: 'www.loremipsum.com',
          tags: { credit: true }
        }
      },
      ids: [1]
    }

    const state = reducer(initial, {
      type: REMOVE_TAG,
      payload: { favoriteId: 1, tagName: 'wallet' }
    })

    expect(state).toEqual(expected)
  })

  it('Should handle FILTER', () => {
    const expected = {
      filterTag: 'wallet',
      entities: {},
      ids: []
    }

    const state = reducer(initialState, {
      type: FILTER,
      payload: { tagName: 'wallet' }
    })

    expect(state).toEqual(expected)
  })

  describe('Selectors', () => {
    it('Should returns ids that contains tag with selector', () => {
      const state = {
        entities: {
          1: {
            id: 1,
            title: 'Lorem Ipsum 1',
            link: 'www.loremipsum1.com',
            tags: { wallet: true, credit: true }
          },
          2: {
            id: 2,
            title: 'Lorem Ipsum 2',
            link: 'www.loremipsum2.com',
            tags: { credit: true }
          },
          5: {
            id: 5,
            title: 'Lorem Ipsum 3',
            link: 'www.loremipsum3.com',
            tags: { wallet: true }
          }
        },
        ids: [1]
      }

      const expected = [1, 5]
      const ids = searchIdByTag(state, 'wallet')
      expect(ids).toEqual(expected)
    })
  })
})
