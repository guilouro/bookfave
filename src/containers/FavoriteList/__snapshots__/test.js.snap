// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavoriteList /> with data Should render correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px 0;
}

.c0:not(:last-child) {
  border-bottom: 1px solid #e7e8eb;
}

.c1 {
  font-size: 16px;
  font-weight: 300;
  margin: 0;
}

.c2 {
  font-size: 12px;
  color: #04a0ff;
  -webkit-text-decoration: none;
  text-decoration: none;
  margin: 10px 0;
  display: table;
}

.c5 {
  outline: none;
  border: none;
  background: transparent;
  margin: 0;
  padding: 0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-transform: uppercase;
  color: #b0b9bf;
}

.c5 span {
  margin-left: 5px;
}

.c3 {
  min-width: 70px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-transform: uppercase;
  padding: 3px 5px;
  border-radius: 2px;
  margin-right: 10px;
  font-size: 10px;
  background-color: #1e68c1;
  color: #fff;
  cursor: pointer;
}

.c4 {
  outline: none;
  border: none;
  background: transparent;
  margin: 0;
  padding: 0;
  cursor: pointer;
  margin-left: 10px;
  color: #fff;
}

<div
  class="c0"
>
  <div
    class=""
  >
    <h2
      class="c1"
    >
      Lorem Ipsum
    </h2>
    <a
      class="c2"
      href="//www.loremipsum.com"
      target="_blank"
    >
      www.loremipsum.com
    </a>
    <div
      class=""
    >
      <span
        class="c3"
      >
        <span>
          wallet
        </span>
        <button
          class="c4"
        >
          âœ•
        </button>
      </span>
    </div>
  </div>
  <button
    class="c5"
  >
    <svg
      height="16px"
      viewBox="0 0 16 18"
      width="16px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 15c0 .551-.448 1-1 1H5c-.552 0-1-.449-1-1V6h8v9zM8 2c1.103 0 2 .897 2 2H6c0-1.103.897-2 2-2zm7 2h-3c0-2.206-1.794-4-4-4S4 1.794 4 4H1a1 1 0 1 0 0 2h1v9c0 1.654 1.346 3 3 3h6c1.654 0 3-1.346 3-3V6h1a1 1 0 1 0 0-2z"
        fill="currentColor"
      />
    </svg>
    <span>
      Delete
    </span>
  </button>
</div>
`;

exports[`<FavoriteList /> without data Should render correctly 1`] = `
.c0 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #b0b9bf;
  padding-top: 50px;
}

<div
  class="c0"
>
  <svg
    fill="currentColor"
    height="50px"
    viewBox="0 0 475 475.884"
    width="50px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M473.738 328.543A23.656 23.656 0 0 0 450.922 312h-10.48V56a8 8 0 0 0-8-8h-112V8a8 8 0 0 0-8-8h-304a8 8 0 0 0-8 8v416a8 8 0 0 0 8 8h264.32l-4 12.336a23.993 23.993 0 0 0 8.695 26.879 24 24 0 0 0 28.25-.016l41.934-30.543a8.06 8.06 0 0 1 9.511 0l42.012 30.559a23.655 23.655 0 0 0 28.223 0 23.66 23.66 0 0 0 8.718-26.871l-16.062-49.489a7.915 7.915 0 0 1 2.934-8.863l42.09-30.558a23.71 23.71 0 0 0 8.671-26.891zM424.441 64v248h-25.52a7.941 7.941 0 0 1-2.472-.406 8.15 8.15 0 0 1-5.117-5.106L375.25 257.04a24.002 24.002 0 0 0-45.602 0l-9.207 28.266V208h64a8 8 0 0 0 0-16h-64v-32h64a8 8 0 0 0 0-16h-64v-32h64a8 8 0 0 0 0-16h-64V64zm-408 352V16h288v296h-50.48a24.004 24.004 0 0 0-22.879 16.566 24.01 24.01 0 0 0 8.734 26.868l42.075 30.55a7.927 7.927 0 0 1 2.945 8.903L277.969 416zm439.2-73.504l-42.079 30.559a23.87 23.87 0 0 0-8.738 26.777l16.082 49.441a7.999 7.999 0 0 1-12.32 8.973l-42.04-30.582a24.2 24.2 0 0 0-28.265 0l-41.992 30.559a7.999 7.999 0 1 1-12.328-8.965l16.074-49.418a23.883 23.883 0 0 0-8.746-26.817l-42.047-30.52A8.002 8.002 0 0 1 253.961 328h52a22.367 22.367 0 0 0 9.648-2.098l.153-.086.144-.046c.114-.051.192-.145.305-.204a23.754 23.754 0 0 0 12.559-14.132L344.852 262a8.003 8.003 0 0 1 7.597-5.504 8.004 8.004 0 0 1 7.602 5.504l16.086 49.473A24.006 24.006 0 0 0 398.922 328h52a7.995 7.995 0 0 1 7.664 5.52 7.994 7.994 0 0 1-2.938 8.976zm0 0"
    />
    <path
      d="M72.441 80h176a8 8 0 0 0 0-16h-176a8 8 0 0 0 0 16zm0 0M72.441 144h176a8 8 0 0 0 0-16h-176a8 8 0 0 0 0 16zm0 0M72.441 208h176a8 8 0 0 0 0-16h-176a8 8 0 0 0 0 16zm0 0M256.441 264a8 8 0 0 0-8-8h-176a8 8 0 0 0 0 16h176a8 8 0 0 0 8-8zm0 0"
    />
  </svg>
  <p>
    We don't have any favorite
  </p>
</div>
`;
